Sources/NIO/LinuxUring.swift

```diff
@@ -0,0 +1,509 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if SWIFTNIO_LIBURING
+#if os(Linux)
+
+import CNIOLinux
+
+// we stuff the event type into the user data for the sqe together with
+// the fd and sequenceIdentifier to match the events without needing any
+// memory allocations or references.
+
+internal enum CqeEventType : Int {
+    case poll = 1, pollModify, pollDelete // start with 1 to not get zero bit patterns for stdin
+}
+
+internal enum UringError: Error {
+    case loadFailure
+    case uringSetupFailure
+    case uringWaitCqeFailure
+}
+
+internal extension TimeAmount {
+    func kernelTimespec() -> __kernel_timespec {
+        var ts = __kernel_timespec()
+        ts.tv_sec = self.nanoseconds / 1_000_000_000
+        ts.tv_nsec = self.nanoseconds % 1_000_000_000
+        return ts
+    }
+}
+
+internal struct UringEvent {
+    var fd : Int32
+    var pollMask : UInt32
+    var sequenceIdentifier : UInt32
+    init () {
+        self.fd = -1
+        self.pollMask = 0
+        self.sequenceIdentifier = 0
+    }
+}
+
+struct fdEventKey: Hashable {
```

Comment:
is this still necessary after #1801 ?

Reply:
I will need to rework after #1801 yep.

Reply:
Rechecked context, we do still need to merge events internally in the backend (having worked in #1801 now).